core/base/index_set.cpp

```diff
@@ -0,0 +1,162 @@
+GKO_REGISTER_OPERATION(global_to_local, index_set::global_to_local);
+GKO_REGISTER_OPERATION(local_to_global, index_set::local_to_global);
+
+
+}  // namespace index_set
+
+
+template <typename IndexType>
+void IndexSet<IndexType>::populate_subsets(const gko::Array<IndexType>& indices,
+                                           const bool is_sorted)
+{
+    auto exec = this->get_executor();
+    this->num_stored_indices_ = indices.get_num_elems();
+    exec->run(index_set::make_populate_subsets(
+        this->index_space_size_, &indices, &this->subsets_begin_,
+        &this->subsets_end_, &this->superset_cumulative_indices_, is_sorted));
+}
+
+
+template <typename IndexType>
+bool IndexSet<IndexType>::contains(const IndexType input_index) const
+{
+    auto local_index = this->get_local_index(input_index);
+    return local_index != invalid_index<IndexType>();
+}
+
+
+template <typename IndexType>
+IndexType IndexSet<IndexType>::get_global_index(const IndexType index) const
+{
+    auto exec = this->get_executor();
+    const auto local_idx =
+        Array<IndexType>(exec, std::initializer_list<IndexType>{index});
+    auto global_idx =
+        Array<IndexType>(exec, this->map_local_to_global(local_idx, true));
+
+    return exec->copy_val_to_host(global_idx.get_data());
+}
+
+
+template <typename IndexType>
+IndexType IndexSet<IndexType>::get_local_index(const IndexType index) const
+{
+    auto exec = this->get_executor();
+    const auto global_idx =
+        Array<IndexType>(exec, std::initializer_list<IndexType>{index});
+    auto local_idx =
+        Array<IndexType>(exec, this->map_global_to_local(global_idx, true));
+
+    return exec->copy_val_to_host(local_idx.get_data());
+}
+
+
+template <typename IndexType>
+Array<IndexType> IndexSet<IndexType>::to_global_indices() const
+{
+    auto exec = this->get_executor();
+    auto num_elems = exec->copy_val_to_host(
+        this->superset_cumulative_indices_.get_const_data() +
+        this->superset_cumulative_indices_.get_num_elems() - 1);
+    auto decomp_indices = gko::Array<IndexType>(exec, num_elems);
+    exec->run(index_set::make_to_global_indices(
+        this->index_space_size_, &this->subsets_begin_, &this->subsets_end_,
+        &this->superset_cumulative_indices_, &decomp_indices));
+
+    return std::move(decomp_indices);
+}
+
+
+template <typename IndexType>
+Array<IndexType> IndexSet<IndexType>::map_local_to_global(
+    const Array<IndexType>& local_indices, const bool is_sorted) const
+{
+    auto exec = this->get_executor();
+    auto global_indices =
+        gko::Array<IndexType>(exec, local_indices.get_num_elems());
+
+    GKO_ASSERT(this->get_num_subsets() >= 1);
+    exec->run(index_set::make_local_to_global(
+        this->index_space_size_, &this->subsets_begin_, &this->subsets_end_,
+        &this->superset_cumulative_indices_, &local_indices, &global_indices,
+        is_sorted));
+    return std::move(global_indices);
```

Comment:
nit: std::move makes copy elision/return value optimization more difficult for the compiler
```suggestion
    return global_indices;
```